### PR TITLE
Add support for inverting CircularNotchedRectangle to optionally be drawn on the bottom of a path

### PR DESCRIPTION
**Pull request by arbaker2**
This PR allows for an optional argument [inverted[ to be passed to the [getOuterPath] method of a CircularNotchedRectangle object in order to invert the notch for situations where it is desired to draw the notch on the bottom of the path. This allows both of the below paths in the below screenshot to be drawn and changes no default behavior. 

Simulator Screenshot - iPhone 15 - 2024-07-07 at 11 46 12

*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*
This add a feature similar to the one discussed in #49973, original feature proposal #151381

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/painting/notched_shapes.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/painting/notched_shapes.dart:58]: small wording suggestion here:

```suggestion
  /// If [inverted] is true, the notch is placed at the bottom of the
  /// rectangle.
```

**Review by nate-thegrate — APPROVED** [review]
There's a little bit that I don't like about this `getOuterPath()` method, but it's not related to the changes you've made here.

LGTM, thanks for the fantastic contribution!

**Comment by nate-thegrate** [comment]
We'll be able to merge this pull request once it gets a second approval.

No guarantees, but hopefully it'll be within a few days!

**Comment by arbaker2** [comment]
> We'll be able to merge this pull request once it gets a second approval.
> 
> No guarantees, but hopefully it'll be within a few days!

Great! Thanks so much Nate! And I appreciate the update to the comments. Your update was much more clear and concise.

**Comment by bleroux** [comment]
Thanks for this good quality PR!

I don't have access to the Google testing failures but It might be related to the added parameter because it will break `CircularNotchedRectangle` subclasses which override `getOuterPath`.
I found this related discussion: https://github.com/dart-lang/sdk/issues/36027.

The PR description points to https://github.com/flutter/flutter/issues/49973, can you explain how one can use this PR to get the shape proposed in https://github.com/flutter/flutter/issues/49973?

**Comment by arbaker2** [comment]
> Thanks for this good quality PR!
> 
> I don't have access to the Google testing failures but It might be related to the added parameter because it will break `CircularNotchedRectangle` subclasses which override `getOuterPath`. I found this related discussion: [dart-lang/sdk#36027](https://github.com/dart-lang/sdk/issues/36027).
> 
> The PR description points to #49973, can you explain how one can use this PR to get the shape proposed in #49973?

It is strange to me that this test failed as the PR with 1 commit passed all tests. The only change in the second commit was a change in comments recommended by @nate-thegrate. There is not much detail on the test as to why it failed. Is there a way to re-run the checks to verify there is an issue?

Also, this PR does not allow for the shape in #49973. I originally opened up an issue #151381 which was closed because it was too similar to #49973. So that is why this PR is linked to this issue.

**Comment by bleroux** [comment]
> It is strange to me that this test failed as the PR with 1 commit passed all tests. The only change in the second commit was a change in comments recommended by @nate-thegrate. There is not much detail on the test as to why it failed. Is there a way to re-run the checks to verify there is an issue?

I agree this is confusing. FYI, this is not related to your second commit.
When a PR is filed the 'Google testing' step is not run. It is run once there is one approval (I don't remember if there is a delay here).

To re-run the check I clicked on 'Details' and then 'Re-run' (I dont know if one have to be a team member to do this). Pushing an empty commit or using the Github 'Update with rebase' button are ways to trigger the whole checks.

**Comment by bleroux** [comment]
Great the check is green now.

@goderbauer
Are we ok with a change that might break subclasses?
This PR adds an optional parameter to `CircularNotchedRectangle.getOuterPath`. There are no subclasses that override this method in the framework and in the customer tests, so can we assume this is not a breaking change?

**Comment by nate-thegrate** [comment]
@bleroux thanks for bringing up the concern about breaking subclasses; it didn't even cross my mind till now.

It looks like there isn't any benefit to subclassing `CircularNotchedRectangle` directly rather than just making another subclass of `NotchedShape`.

<br>

That being said, there is a pretty straightforward way to mitigate subclass breakages:

```dart
class CircularNotchedRectangle extends NotchedShape {
  const CircularNotchedRectangle({this.inverted = false});

  final bool inverted;

  @override
  Path getOuterPath(Rect host, Rect? guest) {
    // ...
  }
}
```

Perhaps it'd be best to implement this way, since `AutomaticNotchedShape` also uses class members rather than optional method arguments.

**Comment by arbaker2** [comment]
@nate-thegrate @bleroux I am aligned with this adjustment as well. Thanks for bringing up these helpful concerns and suggested improvements! I can work to add another commit with this implemented.

**Comment by arbaker2** [comment]
@nate-thegrate @bleroux let me know if these changes look good to you or if further updates are needed. Thanks so much!

**Review by bleroux — APPROVED** [review]
LGTM!
Thanks for this quality PR and for your reactivity 🙏

**Comment by arbaker2** [comment]
@nate-thegrate @bleroux thank you both for your help reviewing this PR!

**Pull request merged**
